kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -176,4 +179,96 @@ public static String getPhysicalName(StructField field) {
+    }
+
+    public static boolean isColumnMappingModeEnabled(String columnMappingMode) {
+        return NAME_OR_ID.contains(columnMappingMode);
+    }
+
+    public static Metadata updateColumnMappingMetadata(Metadata metadata,
+                                                       String columnMappingMode,
+                                                       boolean isNewTable) {
+        switch (columnMappingMode) {
+            case COLUMN_MAPPING_MODE_NONE:
+                return metadata;
+            case COLUMN_MAPPING_MODE_ID: // fall through
+            case COLUMN_MAPPING_MODE_NAME:
+                return assignColumnIdAndPhysicalName(metadata, isNewTable);
+            default:
+                throw new UnsupportedOperationException(
+                        "Unsupported column mapping mode: " + columnMappingMode);
+        }
+    }
+
+    private static Metadata assignColumnIdAndPhysicalName(Metadata metadata, boolean isNewTable) {
+        StructType schema = metadata.getSchema();
+        // TODO: read max col id from delta.columnMapping.maxColumnId
+        int maxColumnId = findMaxColumnId(schema);
+        StructType newSchema = new StructType();
+        for (StructField field : schema.fields()) {
+            StructField updatedField = field;
+
+            if (!hasColumnId(field)) {
+                maxColumnId++;
+                updatedField = field.withNewMetadata(FieldMetadata.builder()
+                        .fromMetadata(field.getMetadata())
+                        .putLong(COLUMN_MAPPING_ID_KEY, maxColumnId)
+                        .build());
+            }
+            if (!hasPhysicalName(field)) {
+                // re-use old display names as physical names when a table is updated
+                String physicalName = isNewTable ? "col-" + UUID.randomUUID() : field.getName();
+                updatedField = updatedField.withNewMetadata(FieldMetadata.builder()
+                        .fromMetadata(updatedField.getMetadata())
+                        .putString(COLUMN_MAPPING_PHYSICAL_NAME_KEY, physicalName)
+                        .build());
+            }
+            newSchema = newSchema.add(updatedField);
+        }
+
+        return metadata.withNewSchema(newSchema);
+    }
+
+    static int findMaxColumnId(StructType schema) {
+        return schema.fields().stream()
+                .filter(ColumnMapping::hasColumnId)
+                .map(ColumnMapping::getColumnId)
+                .max(Integer::compareTo).orElse(0);
+    }
+
+    private static boolean hasColumnId(StructField field) {
+        return field.getMetadata().contains(COLUMN_MAPPING_ID_KEY);
+    }
+
+    private static boolean hasPhysicalName(StructField field) {
+        return field.getMetadata().contains(COLUMN_MAPPING_PHYSICAL_NAME_KEY);
+    }
+
+    private static int getColumnId(StructField field) {
+        // TODO: is there a better way to read this as an int?
```

Comment:
delta-spark's `Metadata` class has some helper methods to get values from a `Map<String, Object>` map in the given type, such as `metadata.getLong(<key>)`.
It could make sense to add something similar to either `FieldMetadata` or to a util class

Reply:
Adding that to the fieldMetadata makes sense. What do you think @vkorukanti ?

@nastra - after Venki responds, you could (1) create a GitHub issue to add those fields, and then (2) update this TODO to reference that issue so we come back to it later

Reply:
yep I'll make sure to follow up on this